### PR TITLE
Login backend

### PR DESCRIPTION
**Pull request by hht-hannah**
Add Login backend. Add HttpManager to set Authorization header and handle get and post requests. 
I also add a drawer to put the logout button. 
Screenshot_20200812_223619_com example Graminsta
Screenshot_20200812_223546_com example Graminsta

### CODE REVIEW

File: flutter/Graminsta/lib/auth/login.dart [hunk 1]
```diff
@@ -34,17 +71,19 @@ class _LoginPageState extends State<LoginPage> {
...
                 decoration: InputDecoration(
                   icon: Icon(
-                    Icons.email,
+                    Icons.person,
                   ),
-                  hintText: "Email Address",
+                  hintText: "Username",
                 ),
                 style: TextStyle(fontSize: 16, color: Colors.black),
               ),
```
Comment (shiyi001) [inline flutter/Graminsta/lib/auth/login.dart]: Maybe we need to add a `validator` to ensure the field is not null? 
https://flutter.dev/docs/cookbook/forms/validation

File: flutter/Graminsta/lib/auth/login.dart [hunk 2]
```diff
@@ -1,11 +1,46 @@
...
 
 class _LoginPageState extends State<LoginPage> {
+  final _usernameController = TextEditingController();
+  final _passwordController = TextEditingController();
+
+  void login() async {
+    Response response;
+    HttpManager.post("core/login/", data: {
+      "username": _usernameController.text,
+      "password": _passwordController.text,
```
Comment (shiyi001) [inline flutter/Graminsta/lib/auth/login.dart]: The password should be hashed for security consideration.
https://stackoverflow.com/questions/47870642/dart-md5-from-string

File: flutter/Graminsta/lib/http.dart [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import 'package:dio/dio.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+//A class to handle post and get request.
+class HttpManager {
+  static String baseUrl = "http://192.168.43.29:8000/";//change into your base url
```
Comment (shiyi001) [inline flutter/Graminsta/lib/http.dart]: Store the `ip` and `port` on `config` file?

File: flutter/Graminsta/lib/auth/login.dart [hunk 4]
```diff
@@ -1,11 +1,46 @@
...
+    if (setTokenResult) {
+      //if set token success, redirect to homepage
+      Navigator.of(context).pushNamedAndRemoveUntil(
+        '/home',
+        (route) => route == null,
+      );
+    } else {
+      debugPrint('保存登录token失败');
+    }
+  }
```
Comment (shiyi001) [inline flutter/Graminsta/lib/auth/login.dart]: Put this function in a separate file，because it will also be used during registration.

File: flutter/Graminsta/lib/auth/login.dart (hunk 2 above)
Comment (Time1ess) [reply flutter/Graminsta/lib/auth/login.dart]: Actually, I think you mean encryption. Yeah, that's a technique to prevent password intercepted, but by default, Django expects real password instead of encrypted one, the hashing of the password is computed during the Django authentication process, for different password, the `salt` and the `hashing algorithm` might be different. And back to this, I think the best thing you can do is let client (Flutter app in this case) to encrypt the password with some public key, and server decrypts the encrypted password with its private key, but it's kind of overkill for a starter project. Usually, if you are using HTTPS, it should provide enough security for this case, as the encryption and decryption is done by SSL. So I'd suggest we just leave this as it is now.

File: flutter/Graminsta/lib/auth/login.dart [hunk 5]
```diff
@@ -1,11 +1,46 @@
...
+      Scaffold.of(context).showSnackBar(SnackBar(
+          content: Text('登录失败，请检查用户名和密码是否正确'),
+          duration: Duration(milliseconds: 2000)));
+      print(e);
+    });
+  }
+
+  _setToken(token) async {
+    final prefs = await SharedPreferences.getInstance();
+    final setTokenResult = await prefs.setString('user_token', token);
```
Comment (Time1ess) [inline flutter/Graminsta/lib/auth/login.dart]: Creates some `constants.dart` file to keep your keys like `user_token`.

```
const userTokenKey = 'user_token'
```

So we don't need to use `'user_token'` everywhere, this can provide us the flexibility to rename the key.

File: backend/graminsta/core/services.py [hunk 6]
```diff
@@ -33,3 +35,34 @@ def create_userinfo(validated_data):
...
+    """User Authentication
+
+    Parameters
+    ----------
+    request: rest_framework.request.Request
+        Request from the frontend
+
+    Returns
+    -------
+    token.key: String format
```
Comment (Time1ess) [inline backend/graminsta/core/services.py]: I think you can return token instead.

File: backend/graminsta/core/services.py [hunk 7]
```diff
@@ -33,3 +35,34 @@ def create_userinfo(validated_data):
         gender=validated_data.pop('gender')
     )
     return user_info
+
+
+def user_authentication(request):
```
Comment (Time1ess) [inline backend/graminsta/core/services.py]: Usually, the name of the method should be some verb + noun, in your case, I think you can just use `create_authentication_token`, since `request` is a concept between client-server, you don't need to know that in your service APIs, you can pass in the `username` and `password`, return a token.

File: backend/graminsta/core/services.py [hunk 8]
```diff
@@ -33,3 +35,34 @@ def create_userinfo(validated_data):
...
+        if the authentication passed.
+    None if the authentication failed.
+    """
+    data = request.data
+    user = authenticate(
+        request,
+        username=data.get("username", ""),
+        password=data.get("password", ""))
+    if user is not None:
+        login(request, user)
```
Comment (Time1ess) [inline backend/graminsta/core/services.py]: I think `login` is to add some instructions to let client to set cookie for authentication, do we need that?

File: backend/graminsta/core/services.py [hunk 9]
```diff
@@ -33,3 +35,34 @@ def create_userinfo(validated_data):
...
+        try:
+            token = Token.objects.get(user=user)
+            token.delete()
+            token = Token.objects.create(user=user)
+        except Token.DoesNotExist:
+            token = Token.objects.create(user=user)
```
Comment (Time1ess) [inline backend/graminsta/core/services.py]: I think this can be better?

```
try:
  Token.objects.get(user=user).delete()
except Token.DoesNotExist:
  # It's OK if there is no existing token.
   pass
finally:
   return Token.objects.create(user=user)
```

File: backend/graminsta/core/views.py [hunk 10]
```diff
@@ -33,3 +33,34 @@ def post(self, request):
...
+            return the anth token if the authentication passes.
+        """
+        token = user_authentication(request)
+        if token is None:
+            return Response(
+                "Login Failed",
+                status=status.HTTP_404_NOT_FOUND
+            )
+        return Response(
+            token,
```
Comment (Time1ess) [inline backend/graminsta/core/views.py]: To return the token, consider using a json map format like: `{'token': token}`

File: flutter/Graminsta/lib/auth/login.dart [hunk 11]
```diff
@@ -1,11 +1,46 @@
...
+    HttpManager.post("core/login/", data: {
+      "username": _usernameController.text,
+      "password": _passwordController.text,
+    }, success: (data) {
+      _setToken(data);
+    }, error: (e) {
+      Scaffold.of(context).showSnackBar(SnackBar(
+          content: Text('登录失败，请检查用户名和密码是否正确'),
+          duration: Duration(milliseconds: 2000)));
+      print(e);
```
Comment (Time1ess) [inline flutter/Graminsta/lib/auth/login.dart]: Remove `print`

File: flutter/Graminsta/lib/auth/login.dart [hunk 12]
```diff
@@ -1,11 +1,46 @@
...
   _LoginPageState createState() => new _LoginPageState();
 }
 
 class _LoginPageState extends State<LoginPage> {
+  final _usernameController = TextEditingController();
+  final _passwordController = TextEditingController();
+
+  void login() async {
+    Response response;
+    HttpManager.post("core/login/", data: {
```
Comment (Time1ess) [inline flutter/Graminsta/lib/auth/login.dart]: Could we have another layer of abstraction like `auth_service.dart`. And maybe we can use [http](https://pub.dev/packages/http) internally instead of `HttpManager`, since it's easy and enough for our case?

```
// http_service.dart
// Class that handles HTTP requests.
class _HttpService {
  static _HttpService _instance;

  final globalHeaders = Map<String, string>();

  factory _HttpService() {
    _instance ??= _HttpService._();
    return _instance;
  }

  _HttpService._() {}

  // Set the auth token for authentication.
  Future<void> setAuthToken(String token);
  
  Future<Response> get(String url, {Map<String, String> headers});

  Future<Response> post(String url, {Map<String, String> headers, Object body, Encoding encoding});
  
  ...
}

// Variable to allow other services to use _HttpService easily.
final http = _HttpService();

// auth_service.dart

import 'path/to/http_service.dart';

class AuthService {
  static AuthService _instance;

  factory AuthService() {
    _instance ??= AuthService._();
    return _instance;
  }

  AuthService._() {}

  Future<void> logIn(String username, String password) {
    http.post(xxx)
  }
}
```

See also: https://flutter.dev/docs/cookbook/networking/fetch-data
https://pub.dev/documentation/http/latest/http/http-library.html

File: flutter/Graminsta/lib/http.dart [hunk 13]
```diff
@@ -0,0 +1,100 @@
+import 'package:dio/dio.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+//A class to handle post and get request.
+class HttpManager {
```
Comment (Time1ess) [inline flutter/Graminsta/lib/http.dart]: See  my comments above for `auth_service.dart` and `http_service.dart`

File: backend/graminsta/core/views.py [hunk 14]
```diff
@@ -50,17 +50,17 @@ def post(self, request):
...
+        response: json format
+            return "Login Failed" if authentication failed.
+            return the token if the authentication passed.
         """
-        token = user_authentication(request)
+        token = create_authentication_token(request.data)
         if token is None:
             return Response(
                 "Login Failed",
                 status=status.HTTP_404_NOT_FOUND
```
Comment (Time1ess) [inline backend/graminsta/core/views.py]: Return HTTP_401 instead?

File: flutter/Graminsta/lib/auth_service.dart [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+  factory AuthService() {
+    _instance ??= AuthService._();
+    return _instance;
+  }
+
+  AuthService._();
+
+  static Future<void> logIn(String username, String password,
+      {Function success, Function error}) {
```
Comment (Time1ess) [inline flutter/Graminsta/lib/auth_service.dart]: A comma before the `}` and reformat to help indent.

```
Function error,})
```

File: flutter/Graminsta/lib/auth_service.dart [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    http.post("core/login/",
+        body: {"username": username, "password": password}).then((res) {
```
Comment (Time1ess) [inline flutter/Graminsta/lib/auth_service.dart]: Ditto for comma before `}`

File: flutter/Graminsta/lib/http_service.dart [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+  factory _HttpService() {
+    _instance ??= _HttpService._();
+    return _instance;
+  }
+
+  _HttpService._();
+
+  // Set the auth token for authentication.
+  Future<void> setAuthToken(String token, {Function success}) async {
```
Comment (Time1ess) [inline flutter/Graminsta/lib/http_service.dart]: You can use `Future<boolean>` to indicate login succeed for failed, and your callback functions like `success` then don't have to be passed here and there, then it can be written as:

```
Future<void> _login() async {
  final isLoggedIn = await AuthService.logIn(_usernameController.text, _passwordController.text);
  if (isLoggedIn) {
    Navigator.of(context).pushNamedAndRemoveUntil(...)
  } else {
    Scaffold.of(context).showSnackBar(SnackBar(...));
  }
}
```

File: flutter/Graminsta/lib/auth_service.dart [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+  static AuthService _instance;
+
+  factory AuthService() {
+    _instance ??= AuthService._();
+    return _instance;
+  }
+
+  AuthService._();
+
+  static Future<void> logIn(String username, String password,
```
Comment (Time1ess) [inline flutter/Graminsta/lib/auth_service.dart]: `Future<boolean>` here too.

File: flutter/Graminsta/lib/http_service.dart [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+      if (success != null) {
+        success();
+      }
+    } else {
+      debugPrint('保存登录token失败');
+    }
+  }
+
+  Future<baseHttp.Response> get(String url,
+      {Map<String, String> headers}) async {
```
Comment (Time1ess) [inline flutter/Graminsta/lib/http_service.dart]: Ditto, add comma.

File: flutter/Graminsta/lib/http_service.dart [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+import 'package:flutter/material.dart';
+import 'package:http/http.dart' as baseHttp;
+import 'package:shared_preferences/shared_preferences.dart';
+import 'package:Graminsta/constants.dart';
+
+// Class that handles HTTP requests.
+class _HttpService {
+  static _HttpService _instance;
+
+  final globalHeaders = Map<String, String>();
```
Comment (Time1ess) [inline flutter/Graminsta/lib/http_service.dart]: You can add prefix `_` to make it private.

File: flutter/Graminsta/lib/http_service.dart [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+    _instance ??= _HttpService._();
+    return _instance;
+  }
+
+  _HttpService._();
+
+  // Set the auth token for authentication.
+  Future<void> setAuthToken(String token, {Function success}) async {
+    final prefs = await SharedPreferences.getInstance();
+    final setTokenResult = await prefs.setString(userTokenKey, token);
```
Comment (Time1ess) [inline flutter/Graminsta/lib/http_service.dart]: Why don't you add this token into the `globalHeaders` map?

File: flutter/Graminsta/lib/http_service.dart [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  }
+
+  Future<baseHttp.Response> get(String url,
+      {Map<String, String> headers}) async {
+    baseHttp.Response response;
+    Map<String, dynamic> headersMap = headers == null ? new Map() : headers;
+    url = baseUrl + url;
+    final prefs = await SharedPreferences.getInstance();
+    final token = prefs.getString(userTokenKey) ?? '';
```
Comment (Time1ess) [inline flutter/Graminsta/lib/http_service.dart]: You don't need to retrieve the token every time you make a request. Instead, keep the token in your `globalHeaders` map, every time you make a request, merge everything from `globalHeaders` to the request's headers. Regarding adding the token to the `globalHeaders`, IIUC, there are only two places:

1. When user authenticated successfully, add the token to `globalHeaders`.
2. When the app starts, try to load any token stored in the SharedPreferences and verify if it's valid by making some requests to server, if it's valid, add it to `globalHeaders`

File: flutter/Graminsta/lib/http_service.dart (hunk 19 above)
Comment (Time1ess) [inline flutter/Graminsta/lib/http_service.dart]: You can rewrite to something like this.
```
Future<baseHttp.Response> get(
    String url, {
    Map<String, String> headers = const {},
}) async {
    headers.addAll(globalHeaders);
    http.Response response;
    try {
        await http.get(`$baseUrl$url`, headers: headers);
    } catch (e) {
        ...
    }
    return response;
}

File: flutter/Graminsta/lib/http_service.dart [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  Future<baseHttp.Response> post(String url,
+      {Map<String, String> headers, Object body, Encoding encoding}) async {
+    baseHttp.Response response;
+    Map<String, String> headersMap = headers == null ? new Map() : headers;
+    Map<String, String> dataMap = body == null ? new Map() : body;
+    url = baseUrl + url;
+    final prefs = await SharedPreferences.getInstance();
+    final token = prefs.getString(userTokenKey) ?? '';
```
Comment (Time1ess) [inline flutter/Graminsta/lib/http_service.dart]: Ditto

**Review by Time1ess — CHANGES_REQUESTED** [review]

File: flutter/Graminsta/lib/http_service.dart (hunk 19 above)
Comment (hht-hannah) [reply flutter/Graminsta/lib/http_service.dart]: I think the header default value cannot be set as const{}, or else it will be an unmodifiable Map.

Comment (Time1ess) [reply flutter/Graminsta/lib/http_service.dart]: No, to provide default value for some container types like `map` and `list` must be const.

> You can also pass lists or maps as default values. The following example defines a function, doStuff(), that specifies a default list for the list parameter and a default map for the gifts parameter.

 See also: https://dart.dev/guides/language/language-tour#optional-parameters

File: flutter/Graminsta/lib/main.dart [hunk 24]
```diff
@@ -41,21 +48,57 @@ class _HomePageState extends State<HomePage> {
...
+    //check user login status. If there is no token, redirect to AuthPage.
+    getUserToken().then((token) => {
+          if (token == "")
+            {
+              Navigator.of(context).pushNamedAndRemoveUntil(
+                '/auth',
+                (route) => route == null,
+              )
+            }
+        });
```
Comment (Time1ess) [inline flutter/Graminsta/lib/main.dart]: Use FutureBuilder instead, since this process is `async`. But I'd suggest you do that in another PR to keep this PR small.

See also: https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html

**Review by Time1ess — CHANGES_REQUESTED** [review]

**Review by Time1ess — APPROVED** [review]
Please update to use FutureBuilder before submitting this PR.

File: flutter/Graminsta/lib/main.dart (hunk 24 above)
Comment (hht-hannah) [reply flutter/Graminsta/lib/main.dart]: I will open another PR to fix this.

File: flutter/Graminsta/lib/serviece/http_service.dart [hunk 25]
```diff
@@ -57,18 +57,19 @@ class _HttpService {
   Future<baseHttp.Response> post(
     String url, {
     Map<String, String> headers,
-    Object body,
+    Object body = const {},
     Encoding encoding,
   }) async {
     baseHttp.Response response;
     Map<String, String> headersMap = headers == null ? new Map() : headers;
```
Comment (Time1ess) [inline flutter/Graminsta/lib/serviece/http_service.dart]: Can be simplified into

```
headers ??= <String, String>{};
```

Comment (Time1ess) [reply flutter/Graminsta/lib/serviece/http_service.dart]: Starting from Dart 2.0, no need for `new`.

**Review by Time1ess — APPROVED** [review]

**Review by shiyi001 — APPROVED** [review]
Great!
